### PR TITLE
SC-1857: Mutagen

### PR DESCRIPTION
**Pull request by zyuzka**
Ticket: https://spryker.atlassian.net/browse/SC-1857
RG: https://release.spryker.com/release-groups/view/2383

### CODE REVIEW

File: bin/mount/mutagen.sh [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        processes+=($!)
+    done
+
+    for i in "${processes[@]}"; do
+        while kill -0 $i >/dev/null 2>&1; do
+            sleep 1
+        done
+    done
+
+    kill -9 ${progressPID}
```
Comment (zyuzka) [inline bin/mount/mutagen.sh]: Kill process always shows message about killing process.
```
>/dev/null 2>&1
```
doesn't help =(

File: bin/mount/mutagen.sh [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    export DOCKER_SYNC_SKIP_UPDATE=1
+    local syncConf="${DEPLOYMENT_PATH}/mutagen.yml"
+
+    case $1 in
+        create)
+            verbose "${INFO}Creating 'data-sync' volume and sync container${NC}"
+            docker volume create --name=spryker_dev_data_sync
+            if [ ! "$(docker ps -a | grep spryker_mutagen_sync_1)" ]; then
+                docker run -d \
+                   --name spryker_mutagen_sync_1 \
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: Please use project namespace

Comment (sprymiker) [reply bin/mount/mutagen.sh]: And for volume below

File: bin/mount/mutagen.sh [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+function sync()
+{
+    export DOCKER_SYNC_SKIP_UPDATE=1
+    local syncConf="${DEPLOYMENT_PATH}/mutagen.yml"
+
+    case $1 in
+        create)
+            verbose "${INFO}Creating 'data-sync' volume and sync container${NC}"
+            docker volume create --name=spryker_dev_data_sync
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: Please use project namespace

File: bin/mount/mutagen.sh [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    kill -13 ${progressPID} > /dev/null 2>&1 || true
+    echo " "
+    echo 'All data was synced.'
+}
+
+function sync()
+{
+    export DOCKER_SYNC_SKIP_UPDATE=1
+    local syncConf="${DEPLOYMENT_PATH}/mutagen.yml"
+
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: Please, introduce `volumeName` (based on project namespace, tag) variable to use everywhere below.

File: generator/index.php [hunk 5]
```diff
@@ -11,6 +11,7 @@
 $deploymentDir = getenv('SPRYKER_DOCKER_SDK_DEPLOYMENT_DIR') ?: '/tmp';
 $projectYaml = getenv('SPRYKER_DOCKER_SDK_PROJECT_YAML') ?: '';
 $projectName = getenv('SPRYKER_DOCKER_SDK_PROJECT_NAME') ?: '';
+$projectSourceDir = getenv('SPRYKER_DOCKER_SDK_PROJECT_SOURCE_DIR') ?: __DIR__ ;
```
Comment (sprymiker) [inline generator/index.php]: This should not be used during bootstrap.

File: generator/src/templates/mutagen.yml.twig [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        - '.npm'
+        - 'mutagen.*'
+        - 'docker-compose.*'
+        - '*.env'
+        - 'bin/'
+        - 'context/'
+        - 'env/'
+        - 'images/'
+  project:
+    alpha: "{{ _projectSourceDir }}"
```
Comment (sprymiker) [inline generator/src/templates/mutagen.yml.twig]: Please, consider a relative path. Thanks

File: bin/mount/mutagen.sh [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            mutagen project resume ${syncConf}
+            ;;
+
+        clean)
+            mutagen project flush ${syncConf}
+            docker volume rm ${SPRYKER_DOCKER_PREFIX}_${SPRYKER_DOCKER_TAG}_data_sync || true
+            ;;
+
+        stop)
+            mutagen project terminate ${syncConf}
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: It seems docker container is not stopped here. Please make sure it is stopped.

File: bin/mount/mutagen.sh [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                   nc -l 9000
+            fi
+            ;;
+
+        recreate)
+            mutagen project resume ${syncConf}
+            ;;
+
+        clean)
+            mutagen project flush ${syncConf}
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: It seems docker container is not stopped here. Please make sure it is stopped.

File: bin/mount/mutagen.sh [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            docker volume rm ${SPRYKER_DOCKER_PREFIX}_${SPRYKER_DOCKER_TAG}_data_sync || true
+            ;;
+
+        stop)
+            mutagen project terminate ${syncConf}
+            ;;
+        *)
+            if [ $(docker ps | grep 5000 | grep ${SPRYKER_DOCKER_PREFIX}_${SPRYKER_DOCKER_TAG}_data_sync | wc -l |sed 's/^ *//') -eq 0 ]; then
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: 5000 port is used by docker-sync. Is it used for mutagen too? Please check that this IF works as expected. Thanks

File: bin/boot-deployment.sh [hunk 10]
```diff
@@ -57,6 +57,7 @@ function bootDeployment()
         -e SPRYKER_DOCKER_SDK_PROJECT_YAML="/data/deployment/project.yml" \
         -e SPRYKER_DOCKER_SDK_DEPLOYMENT_DIR="/data/deployment" \
         -e SPRYKER_DOCKER_SDK_PLATFORM="$(getPlatform)" \
+        -e SPRYKER_DOCKER_SDK_PROJECT_SOURCE_DIR="${PROJECT_DIR}" \
```
Comment (sprymiker) [inline bin/boot-deployment.sh]: Please, rely on a relative path.

File: bin/mount/mutagen.sh [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    maxCount=3
+
+    while true; do
+        command=$(mutagen sync list ${projectName} | grep Status:)
+
+        if [ "${count}" == "${maxCount}" ]; then
+            break;
+        fi
+
+        if [ ! "${command}" == "${waitingStatus}" ]; then
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: this can be ELSE for the next IF

File: bin/mount/mutagen.sh [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            count=0
+        fi
+
+        if [ "${command}" == "${waitingStatus}" ]; then
+            ((count++))
```
Comment (sprymiker) [inline bin/mount/mutagen.sh]: count=$((count + 1))

**Comment by sprymiker** [comment]
It seems SC-1854 is not proper ticket for this

File: bin/mount/mutagen.sh (hunk 7 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: stopCliContainer() stops all CLI containers

File: bin/mount/mutagen.sh (hunk 8 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: same

File: bin/mount/mutagen.sh (hunk 9 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: Changed

File: bin/boot-deployment.sh (hunk 10 above)
Comment (zyuzka) [reply bin/boot-deployment.sh]: deleted

File: bin/mount/mutagen.sh (hunk 7 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: Added additional function for stoping sync container

File: bin/mount/mutagen.sh (hunk 2 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: fixed

File: bin/mount/mutagen.sh (hunk 3 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: fixed

File: bin/mount/mutagen.sh (hunk 4 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: fixed

File: generator/index.php (hunk 5 above)
Comment (zyuzka) [reply generator/index.php]: deleted

File: generator/src/templates/mutagen.yml.twig (hunk 6 above)
Comment (zyuzka) [reply generator/src/templates/mutagen.yml.twig]: fixed

File: bin/mount/mutagen.sh (hunk 11 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: fixed

File: bin/mount/mutagen.sh (hunk 12 above)
Comment (zyuzka) [reply bin/mount/mutagen.sh]: fixed

**Comment by alexanderM91** [comment]
Hi, @zyuzka, please prepare RG.

**Comment by zyuzka** [comment]
> Hi, @zyuzka, please prepare RG.

Hi=) Updated

**Comment by sprymiker** [comment]
For those who are interested in this topic, please, read https://github.com/spryker/docker-sdk/issues/122#issuecomment-655762948.

**Comment by alexanderM91** [comment]
It's not relevant anymore.
Mutagen support has been released https://github.com/spryker/docker-sdk/releases/tag/1.19.0

**Comment by sprymiker** [comment]
Closed.
Covered by https://github.com/spryker/docker-sdk/pull/136.

**Pull request closed without merging**